src/Tools/dotnet-format/CodeFormatter.cs

```diff
@@ -0,0 +1,131 @@
+using Roslyn.Utilities;
+using CSharpSyntaxKind = Microsoft.CodeAnalysis.CSharp.SyntaxKind;
+using VisualBasicSyntaxKind = Microsoft.CodeAnalysis.VisualBasic.SyntaxKind;
+using Microsoft.CodeAnalysis.Tools.CodeFormatter.MSBuild;
+using System.Diagnostics;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.CodeAnalysis.Tools.CodeFormatter
+{
+    internal class CodeFormatter
+    {
+        public static async Task<int> FormatWorkspaceAsync(ILogger logger, string workspacePath, bool isSolution, CancellationToken token)
+        {
+            logger.LogInformation(string.Format(Resources.Info_FormattingFilesInWorkspace, Path.GetFileName(workspacePath)));
+
+            logger.LogTrace($"Loading workspace.");
+
+            var stopwatch = Stopwatch.StartNew();
+
+            var properties = new Dictionary<string, string>
+            {
+                // This property ensures that XAML files will be compiled in the current AppDomain
+                // rather than a separate one. Any tasks isolated in AppDomains or tasks that create
+                // AppDomains will likely not work due to https://github.com/Microsoft/MSBuildLocator/issues/16.
+                { "AlwaysCompileMarkupFilesInSeparateDomain", bool.FalseString },
+
+                // Use the latest language version to force the full set of available analyzers to run on the project.
+                { "LangVersion", "latest" },
+            };
+            using (var workspace = MSBuildWorkspace.Create(properties))
+            {
+                var projectPath = string.Empty;
+                if (isSolution)
+                {
+                    await workspace.OpenSolutionAsync(workspacePath, cancellationToken: token).ConfigureAwait(false);
+                }
+                else
+                {
+                    await workspace.OpenProjectAsync(workspacePath, cancellationToken: token).ConfigureAwait(false);
+                    projectPath = workspacePath;
+                }
+
+                logger.LogTrace($"Workspace loaded in {stopwatch.ElapsedMilliseconds}ms.");
+                stopwatch.Restart();
+
+                await FormatFilesInSolutionAsync(logger, workspace.CurrentSolution, projectPath, token).ConfigureAwait(false);
+
+                logger.LogTrace($"Format complete in {stopwatch.ElapsedMilliseconds}ms.");
+            }
+
+            logger.LogInformation(Resources.Info_FormatComplete);
+
+            return 0;
+        }
+
+        private static async Task FormatFilesInSolutionAsync(ILogger logger, Solution solution, string projectPath, CancellationToken cancellationToken)
+        {
+            foreach (var projectId in solution.ProjectIds)
+            {
+                var project = solution.GetProject(projectId);
+                if (!string.IsNullOrEmpty(projectPath) && !project.FilePath.Equals(projectPath, StringComparison.OrdinalIgnoreCase))
+                {
+                    continue;
```

Comment:
should we log this case?

Reply:
>  guess, as discussed earlier, some renamings could change files in related projects already. I am happy to hear if anyone has thoughts one way or the other.

Right now i don't see anything related to renaming in your PR.  It's only calling hte formatter.   This is purely syntactic, and only works a file at a time.